src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py

```diff
@@ -256,6 +259,20 @@
                      'filename': offer_interactions_filename}}
 
 
+def get_all_product_categories():
+    service_url = os.environ['ProductsServiceExternalUrl']
+    all_categories_url = service_url + '/categories/all'
+    logger.info(f"Getting full list of categories from URL {all_categories_url}")
+    response = urllib3.PoolManager().request('GET', all_categories_url, retries=False)
+    if response.status != 200:
+        errmsg = f"{response.status}: Products service not working - reason: {response.reason} response: {response.data}"
+        logger.error(errmsg)
+        raise Exception(errmsg)
+    else:
+        categories = json.loads(response.data)
+        return categories
```

Comment:
There could be a race condition here after the Products service is created in ECS and before the Products service is initially built and deployed. As long as this function is called well into the Personalize pre-create process, it should be okay. Filters should be kept towards the end. Maybe note this with a comment?

Reply:
Yes, the Personalize lambda will fail till products are loaded. We will add the comment.